lib/queues.js

```diff
@@ -611,5 +610,22 @@ module.exports = class Queues {
     // Lazy load configuration.
     return this._ironium.configPromise;
   }
+
+  // Returns the maximum amount of jobs a queue should try to reserve.
+  // Every queue gets a share of reservation capacity to better deal
+  // with load spikes across queues.
+  getReserveCapacity() {
+    return this.configPromise
+      .then(config => {
+        const totalCapacity       = config.concurrency - this._activeJobs;
+        const maxCapacityPerQueue = Math.floor(totalCapacity / this._queues.size);
+        const capacity            = Math.min(totalCapacity, maxCapacityPerQueue);
```

Comment:
This is the "reserve" capacity – how many jobs to wait for on each reserve call, at the most.

The goal with this (flawed) algorithm was to better handle spikes of workload.

If all queues are allowed to reserve up to n jobs where n = concurrency, then a spike of n jobs on each queue would cause actual concurrency to be concurrency * queue count.

So we allow each queue to reserve a fair share of jobs each time. This is different from running said amount of jobs concurrently. `_processContinuously` will call this function and as long as there's still capacity to reserve, it will continue consuming jobs.

This is under the assumption that we don't want to reserve jobs that we're not yet ready to process.